# Table of Contents
1. Import libraries
2. Import data: products.csv, orders_wrangled.csv
3. Data cleaning on products.csv: missing values and duplicate checks
4. Export cleaned products: products_checked2.csv
5. Data cleaning on orders_wrangled.csv: drop Unnamed column, missing values, mixed data types, duplicates
6. Export cleaned orders: orders_checked.csv

# Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# Import Data

In [2]:
# Define path
path = r'/users/stephenhelvig/documents/python projects/instacart basket analysis'

In [3]:
# Import products.csv
df_products = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
# Import orders_wrangled.csv
df_orders = pd.read_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# Products.csv data wrangling

In [7]:
# Look for missing values in df.products
df_products.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [8]:
# Find duplicates
df_dups = df_products[df_products.duplicated()]

In [9]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [10]:
df_products.shape

(49693, 5)

In [11]:
# drop duplicates from df_products
df_products_clean_no_dups = df_products.drop_duplicates()

In [12]:
df_products_clean_no_dups.shape

(49688, 5)

In [13]:
# Export cleaned data
df_products_clean_no_dups.to_csv(os.path.join(path, 'Data', 'Prepared Data', 'products_checked2.csv'), index=False)

# Orders.csv - further data cleaning

In [22]:
# Consistency check for df_orders
df_orders.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

Everything looks good here with the exception that 'Unnamed' may just be an index column. The count for days_since_prior_order is lower than the rest of the columns. This likely means that when there is no 2nd order yet, the entry stays blank. 

In [23]:
# check if all the values in 'Unnamed:0' are the same as the index column
(df_orders.index == df_orders['Unnamed: 0']).all()

np.True_

In [24]:
# drop the 'Unnamed: 0' column
df_orders = df_orders.drop(columns=['Unnamed: 0'])

In [25]:
df_orders.describe()

order_id       user_id  order_number  order_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06       3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01       2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01       2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00       0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00       1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01       3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01       5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02       6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [26]:
# check for mixed data types
for col in df_orders.columns.tolist():
	weird = (df_orders[[col]].map(type) != df_orders[[col]].iloc[0].apply(type)).any(axis = 1)
	if len (df_orders[weird]) > 0:
		print(col)

In [27]:
# check for missing values
df_orders.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206,209 missing values in 'days_since_prior_order'. It's possible this column doesn't perform a calculation until another order is placed and therefore, is still empty from customers that have only purchased from Instacart a single time.

In [28]:
# create a new flag column (boolean) to address if the order is a first time order or not
df_orders['first_order'] = df_orders['days_since_prior_order'].isna()

In [29]:
df_orders

order_id  user_id  order_number  order_day_of_week  \
0         2539329        1             1                  2   
1         2398795        1             2                  3   
2          473747        1             3                  3   
3         2254736        1             4                  4   
4          431534        1             5                  4   
...           ...      ...           ...                ...   
3421078   2266710   206209            10                  5   
3421079   1854736   206209            11                  4   
3421080    626363   206209            12                  1   
3421081   2977660   206209            13                  1   
3421082    272231   206209            14                  6   

         order_hour_of_day  days_since_prior_order  first_order  
0                        8                     NaN         True  
1                        7                    15.0        False  
2                       12                    21.0        False  
3                        7                    29.0        False  
4                       15                    28.0        False  
...                    ...                     ...          ...  
3421078                 18                    29.0        False  
3421079                 10                    30.0        False  
3421080                 12                    18.0        False  
3421081                 12                     7.0        False  
3421082                 14                    30.0        False  

[3421083 rows x 7 columns]

I used a flag column to deal with these missing values because removing them or imputing them are not good options. 

In [30]:
# Find duplicates
df_dups = df_orders[df_orders.duplicated()]

No duplicates found

In [34]:
# Exporting cleaned orders_checked.csv
df_orders.to_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_checked.csv'), index = False)